docs/DSP: Add sections on 16-bit and 40-bit modes and on main and extended opcode writing to the same register

Pokechu22 opened the pull request:
Along with various typo fixes in both the DSP manual and comments in Dolphin's DSP code.  The new sections are based on studying Dolphin's DSP interpreter, and not by hardware testing.

This is a draft because I am considering replacing the existing "Perform an additional operation depending on destination register." messages in various instructions with a more explicit back-reference to the "16-bit and 40-bit modes" subsection (along with adding it to any instructions where it's missing but exists in Dolphin), but haven't done so yet.

BhaaLseN commented on docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex:
```diff
@@ -4338,9 +4355,18 @@ \section{Extended opcodes}
 
+Extended opcodes are run \textit{in parallel} with the main opcode; they see the same register state as the input. (For instance, \texttt{\Opcode{MOVR}\Opcode{'MV} \Register{\$ac1}, \Register{\$ax0.l} : \Register{\$ax0.l}, \Register{\$ac1.m}} (encoded as \Value{0x6113}) \textit{swaps} the values of \Register{\$ac1.m} and \Register{\$ax0.l} (and also extends the new value of \Register{\$ac1.m} int ...
+
+Since they are executed in parallel, the main and extension opcodes could theoretically write to the same registers. All opcodes that support extension only modify a main accumulator \Register{\$acD}, as well as \Register{\$prod}, \Register{\$sr}, and/or \Register{\$pc}, while the extension opcodes themselves generally only modify an additional accumulator \Register{\$axD} and addressing register ...
```
Double space, doubt that was intentional
```suggestion
Since they are executed in parallel, the main and extension opcodes could theoretically write to the same registers. All opcodes that support extension only modify a main accumulator \Register{\$acD}, as well as \Register{\$prod}, \Register{\$sr}, and/or \Register{\$pc}, while the extension opcodes themselves generally only modify an additional accumulator \Register{\$axD} and addressing registers \Register{\$arS}. The exception is \Opcode{'L} and \Opcode{'LN}, which has the option of writing to \Register{\$acD}. Thus, \texttt{\Opcode{INC}\Opcode{'L} \Register{\$ac0} : \Register{\$ac0.l}, @\Register{\$ar0}} (encoded as \Value{0x7660}) increments \Register{\$ac0} (and thus \Register{\$ac0.l}), but also sets \Register{\$ac0.l} to the value in data memory at address \Register{\$ar0} and increments \Register{\$ar0}.
```

Pokechu22 replied:
Yep, that space was where TeXworks was wrapping the line for me, so the duplicate space wasn't visible.

BhaaLseN commented on docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex:
```diff
@@ -4338,9 +4355,18 @@ \section{Extended opcodes}
+
+Since they are executed in parallel, the main and extension opcodes could theoretically write to the same registers. All opcodes that support extension only modify a main accumulator \Register{\$acD}, as well as \Register{\$prod}, \Register{\$sr}, and/or \Register{\$pc}, while the extension opcodes themselves generally only modify an additional accumulator \Register{\$axD} and addressing register ...
+
+When the main and extension opcodes write to the same register, the register is set to the two valued bitwise-or'd together. For the above example, \Register{\$ar0.l} would be set to \InlineExpression{(\Register{\$ar0.l} + 1) | MEM[\Register{\$ar0}]}. \textbf{Note that no official uCode writes to the same register twice like this.}
```
I believe you meant "...the two values..."?
```suggestion
When the main and extension opcodes write to the same register, the register is set to the two values bitwise-or'd together. For the above example, \Register{\$ar0.l} would be set to \InlineExpression{(\Register{\$ar0.l} + 1) | MEM[\Register{\$ar0}]}. \textbf{Note that no official uCode writes to the same register twice like this.}
```

BhaaLseN commented on docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex:
```diff
@@ -4338,9 +4355,18 @@ \section{Extended opcodes}
+
+When the main and extension opcodes write to the same register, the register is set to the two valued bitwise-or'd together. For the above example, \Register{\$ar0.l} would be set to \InlineExpression{(\Register{\$ar0.l} + 1) | MEM[\Register{\$ar0}]}. \textbf{Note that no official uCode writes to the same register twice like this.}
+
+Dolphin only supports this multi-write behavior in the DSP Interpreter when the \texttt{PRECISE\_BACKLOG} define is present.  In \texttt{Interpreter::ExecuteInstruction}, if an extended opcode is in use, the extended opcode's behavior is executed first, followed by the main opcode's behavior. The extended opcode does not directly write to registers, but instead records the writes into a backlog ( ...
```
```suggestion
Dolphin only supports this multi-write behavior in the DSP Interpreter when the \texttt{PRECISE\_BACKLOG} define is present. In \texttt{Interpreter::ExecuteInstruction}, if an extended opcode is in use, the extended opcode's behavior is executed first, followed by the main opcode's behavior. The extended opcode does not directly write to registers, but instead records the writes into a backlog (\texttt{WriteToBackLog}). The main opcode calls \texttt{ZeroWriteBackLog} after it is done reading the register values; this directly writes zero to all registers that have pending writes in the backlog. The main opcode then is free to write directly to registers it changes. Afterwards, \texttt{ApplyWriteBackLog} bitwise-ors the value of the register and the value in the backlog; if the main opcode didn't write to the register then \texttt{ZeroWriteBackLog} means that the pending value is being or'd with zero, so it's used without changes. When \texttt{PRECISE\_BACKLOG} is not defined, \texttt{ZeroWriteBackLog} does nothing and \texttt{ApplyWriteBackLog} overwrites the register value with the value from the backlog (so writes from extended opcodes ``win'' over the main opcode).
```

BhaaLseN reviewed (commented):
Seems good for the most part. I'm trying to remember why those opcodes use `r` instead of `d`, but so far I'm coming up empty 🤷

Pokechu22 commented:
Note that there used to be a lot more inconsistencies with the opcode bits (see #10048, particularly 446b1d2f13d98b32db8e0aef59ec29ae06bc2794).  `d` versus `r` is less important; "5.3 Bit meanings" says `d` is destination register while `r` is register (either source or destination), and since `ADD` adds to the value already in the register, `r` could theoretically work, but all of the other `ADD` instructions use `d`.

lioncash commented on docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex:
```diff
@@ -4338,9 +4355,18 @@ \section{Extended opcodes}
+
+When the main and extension opcodes write to the same register, the register is set to the two values bitwise-or'd together. For the above example, \Register{\$ar0.l} would be set to \InlineExpression{(\Register{\$ar0.l} + 1) | MEM[\Register{\$ar0}]}. \textbf{Note that no official uCode writes to the same register twice like this.}
+
+Dolphin only supports this multi-write behavior in the DSP Interpreter when the \texttt{PRECISE\_BACKLOG} define is present. In \texttt{Interpreter::ExecuteInstruction}, if an extended opcode is in use, the extended opcode's behavior is executed first, followed by the main opcode's behavior. The extended opcode does not directly write to registers, but instead records the writes into a backlog (\ ...
```
I don't think we should be including Dolphin specific implementation information in a spec document that's intended to document only the hardware. This should be in the source code somewhere rather than here.

Pokechu22 replied:
I kinda agree, but I feel like it's important to at least mention that Dolphin doesn't emulate it by default and how to enable it, and if that's done explaining how it's implemented is somewhat helpful.

Most of this information is already present in the source code (`WriteToBackLog` in `DSPInterpreter.h` and `ApplyWriteBackLog` and `ZeroWriteBackLog` in `DSPInterperter.cpp`), just without a higher-level overview.

lioncash replied:
I still think documentation to notify people of emulation differences shouldn't be in this document because most people are not going to read hardware documentation to find out about a software difference.

Maybe consider making a top-level `QUIRKS.md` or some similar document in the source tree dedicated for documenting emulation differences (for both intentional and unintentional choices), which gives it way more visibility than in a latex document tucked away inside the `docs/DSP` folder

It also easier to maintain because most people are not going to think about updating the DSP document if things change in the future

Pokechu22 replied:
I'm going to wait for a second opinion on this (while I work on some other DSP-related stuff).

JosJuice replied:
I agree with lioncash.

Pokechu22 replied:
> I still think documentation to notify people of emulation differences shouldn't be in this document because most people are not going to read hardware documentation to find out about a software difference.

My reasoning is that if someone were to see something like `INC’L $ac0 : $ac0.l, @$ar0` in uCode that Dolphin seems to not be handling correctly, it would make sense to check the hardware documentation to determine what's supposed to happen in that case, and the explanation of where and how it's implemented in Dolphin would be useful in that case.

I don't think something like `QUIRKS.md` is a bad idea, but I'm not up for writing something like that right now.  I guess I'll just remove this paragraph for the time being.

lioncash replied:
> I don't think something like QUIRKS.md is a bad idea, but I'm not up for writing something like that right now. I guess I'll just remove this paragraph for the time being.

Sounds good

Pokechu22 replied:
I've moved the paragraph next to the (commented out) `PRECISE_BACKLOG` define, and also moved a few comments so that they're all in DSPInterpreter.cpp (instead of being mixed between that and DSPInterperter.h).

Pokechu22 commented:
I've added hardware tests for the 40-bit mode (one testing a large range of values, though not all of them as that would take several days to run, and one testing various instructions), replaced the cryptic "Perform an additional operation depending on destination register" messages with a reference to the new 40-bit mode section, and also fixed a few more typos. The new version of the manual is here: [GameCube_DSP_Users_Manual.pdf](https://github.com/dolphin-emu/dolphin/files/8809983/GameCube_DSP_Users_Manual.pdf).

Pokechu22 commented:
I discovered that `LOOP` and `BLOOP` also experience saturation; this has been implemented in #10716.  I also realized that I didn't include `'LS` in the test or the list of instructions that experience saturation; that has been fixed (and I also added a note that explains how `'LS` and `'SL` differ because it's not obvious at first glance).
